examples/opt_serving/interactive_hosted.py

```diff
@@ -67,51 +85,126 @@ def batching_loop(timeout=TIMEOUT_MS, max_tokens=MAX_BATCH_TOKENS, max_bs=MAX_BS
+                raise queue.Empty
+            else:
+                # batch is empty or under budget
+                logprobs_batch.append(item)
+    except:
+        logger.debug(f"Prepare to process logprobs batch: {logprobs_batch}")
+        if logprobs_batch:
+            made_model_call = True
+            for work_item in logprobs_batch: # for loop over work items to keep it as batch size 1 for now, to avoid annoying logic with mixing past key values
+                cache_ids = []
                 request_object = {
                     "inputs": [],
                     "min_tokens": [],
                     "max_tokens": [],
                 }
-                for work_item in batch:
-                    ro = work_item.data
-                    request_object["inputs"].append(ro["input"])
-                    request_object["min_tokens"].append(ro.get("min_tokens", 0))
-                    request_object["max_tokens"].append(
-                        ro.get("max_tokens", MAX_SEQ_LEN))
-                    # assumption: everyone has the same remaining args
-                    for key in [
-                            "temperature",
-                            "top_p",
-                            "n",
-                            "best_of",
-                            "echo",
-                            "logprobs",
-                            "stop",
-                    ]:
-                        if key in ro:
-                            request_object[key] = ro[key]
+                ro = work_item.data
+                request_object["inputs"].append(ro["input"])
+                request_object["min_tokens"].append(ro.get("min_tokens", 0))
+                request_object["max_tokens"].append(
+                    ro.get("max_tokens", MAX_SEQ_LEN))
+                cache_ids.append(ro.get("cache_id", None))
+                # assumption: everyone has the same remaining args
+                for key in [
+                        "temperature",
+                        "top_p",
+                        "top_k",
+                        "n",
+                        "best_of",
+                        "echo",
+                        "logprobs",
+                        "stop",
+                ]:
+                    if key in ro:
+                        request_object[key] = ro[key]
                 # do the actual generations
                 request_object["seed"] = random.randint(1, 20000)
-                generations = generator.generate(**request_object)
+                outputs = generator(request_object['inputs'], cache_ids, pasts=PAST_CACHE)
+                generations = []
+                for cache_id, output in zip(cache_ids, outputs):
+                    PAST_CACHE[cache_id] = (time.time(), output.past_key_values) # add to or update the cache with newly computed values
+                    if len(PAST_CACHE) > PAST_CACHE_SIZE_LIMIT: # delete oldest key in cache if cache too big
```

Comment:
What does this cache mean? is it the attention cache or something else?

Reply:
It's a cache for the past_key_values, yeah.